src/sentry/integrations/source_code_management/search.py

```diff
@@ -0,0 +1,110 @@
+from abc import ABC, abstractmethod
+from typing import Any
+
+from django.db.models import Q
+from rest_framework import serializers
+from rest_framework.exceptions import NotFound
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from sentry.api.api_owners import ApiOwner
+from sentry.api.api_publish_status import ApiPublishStatus
+from sentry.api.base import control_silo_endpoint
+from sentry.integrations.api.bases.integration import IntegrationEndpoint
+from sentry.integrations.base import IntegrationInstallation
+from sentry.integrations.mixins.issues import IssueBasicIntegration
+from sentry.integrations.models.integration import Integration
+from sentry.integrations.source_code_management.repository import RepositoryIntegration
+from sentry.organizations.services.organization import RpcOrganization
+
+
+class SourceCodeSearchSerializer(serializers.Serializer):
+    field = serializers.CharField(required=True)
+    query = serializers.CharField(required=True)
+
+
+@control_silo_endpoint
+class SourceCodeSearchEndpoint(IntegrationEndpoint, ABC):
+    owner = ApiOwner.ECOSYSTEM
+    publish_status = {
+        "GET": ApiPublishStatus.PRIVATE,
+    }
+
+    @property
+    def issue_field(self) -> str:
+        return "externalIssue"
+
+    # not used in VSTS
+    @property
+    def repository_field(self) -> str:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def integration_provider(self) -> str | None:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def installation_class(
+        self,
+    ) -> (
+        type[IssueBasicIntegration | RepositoryIntegration]
+        | tuple[type[IssueBasicIntegration | RepositoryIntegration]]
+    ):
+        raise NotImplementedError
+
+    @abstractmethod
+    def handle_search_issues(
+        self, installation: IntegrationInstallation, query: str, repo: str
```

Comment:
i would love to be able to type `installation` as type `self.installation_class`, but i haven't been able to successfully do so

Reply:
Good thinking. Something roughly like this should work:

```python
I = TypeVar("I", bound=IntegrationInstallation)

class SourceCodeSearchEndpoint(IntegrationEndpoint, Generic[I], ABC):
    @property
    @abstractmethod
    def installation_class(self) -> type[I]: ...

    @abstractmethod
    def handle_search_issues(self, installation:I, query: str, repo: str): ...
```

Reply:
that did it! thank you